src/jrd/exe.cpp

```diff
@@ -878,7 +878,14 @@ void EXE_start(thread_db* tdbb, jrd_req* request, jrd_tra* transaction)
 	if (transaction->tra_flags & TRA_prepared)
 		ERR_post(Arg::Gds(isc_req_no_trans));
 
-	JrdStatement* statement = request->getStatement();
+	const auto dbb = tdbb->getDatabase();
+	const auto statement = request->getStatement();
+
+	// Generate request id.
+	request->setRequestId(
+		request->isRequestIdUnassigned() && statement->requests.hasData() && request == statement->requests[0] ?
```

Comment:
This part (`statement->requests.hasData() && request == statement->requests[0]`) is quite longish and repeated twice across the code. Maybe encapsulate it into something like `bool Statement::isRootRequest(jrd_req request) const` or probably `bool jrd_req::isRoot() const`? "Root" could be  replaced with "Initial" or whatever.